pkg/compact/downsample.go

```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		return id, err
+	}
+	defer h.Close()
+
+	q, err := tsdb.NewBlockQuerier(b, b.Meta().MinTime, b.Meta().MaxTime)
+	if err != nil {
+		return id, errors.Wrap(err, "create block querier")
+	}
+	defer q.Close()
+
+	set, err := q.Select(labels.NewEqualMatcher(index.AllPostingsKey()))
+	if err != nil {
+		return id, errors.Wrap(err, "select all series")
+	}
+	for set.Next() {
+		s := set.At()
+
+		app := h.Appender()
+
+		if err := downsampleAggr(app, s.Labels(), s.Iterator(), window); err != nil {
+			return id, errors.Wrap(err, "downsample average")
+		}
+		if err := app.Commit(); err != nil {
+			return id, errors.Wrap(err, "commit")
+		}
+	}
+	if set.Err() != nil {
+		return id, errors.Wrap(set.Err(), "iterate series set")
+	}
+	comp, err := tsdb.NewLeveledCompactor(nil, log.NewNopLogger(), []int64{b.Meta().MaxTime - b.Meta().MinTime}, nil)
+	if err != nil {
+		return id, errors.Wrap(err, "create compactor")
+	}
+	id, err = comp.Write(dir, h, b.Meta().MinTime, b.Meta().MaxTime)
+	if err != nil {
+		return id, errors.Wrap(err, "compact head")
+	}
+	bdir := filepath.Join(dir, id.String())
+
+	origMeta, err := block.ReadMetaFile(b.Dir())
+	if err != nil {
+		return id, errors.Wrap(err, "read block meta")
+	}
+
+	meta, err := block.ReadMetaFile(bdir)
+	if err != nil {
+		return id, errors.Wrap(err, "read block meta")
+	}
+	meta.Thanos.Labels = origMeta.Thanos.Labels
+	meta.Thanos.DownsamplingWindow = window
+	meta.Compaction = origMeta.Compaction
+
+	if err := block.WriteMetaFile(bdir, meta); err != nil {
+		return id, errors.Wrap(err, "write block meta")
+	}
+	return id, nil
+}
+
+type aggregator struct {
+	total   int     // total samples processed
+	count   int     // samples in current window
+	sum     float64 // value sum of current window
+	min     float64 // min of current window
+	max     float64 // max of current window
+	counter float64 // total counter state since beginning
+	resets  int     // number of counter resests since beginning
+	last    float64 // last added value
+}
+
+func (a *aggregator) reset() {
+	a.count = 0
+	a.sum = 0
+	a.min = math.MaxFloat64
+	a.max = -math.MaxFloat64
+}
+
+func (a *aggregator) add(v float64) {
+	if a.total > 0 {
+		if v < a.last {
+			// Counter reset, correct the value.
+			a.counter += v
+			a.resets++
+		} else {
+			// Add delta with last value to the counter.
+			a.counter += v - a.last
+		}
+	} else {
+		// First sample sets the counter.
+		a.counter = v
+	}
+	a.last = v
+
+	a.sum += v
+	a.count++
+	a.total++
+
+	if v < a.min {
+		a.min = v
+	}
+	if v > a.max {
+		a.max = v
+	}
+}
+
+func (a *aggregator) get() float64 {
+	return a.sum / float64(a.count)
+}
+
+func aggrLset(lset labels.Labels, aggr string) labels.Labels {
+	res := make(labels.Labels, len(lset))
+	copy(res, lset)
+
+	for i, l := range res {
+		if l.Name == "__name__" {
+			res[i].Value = fmt.Sprintf("%s$%s", l.Value, aggr)
+		}
+	}
+	return res
+}
+
+func isCounter(lset labels.Labels) bool {
```

Comment:
sooo this is not catching everything for sure.. can we at least add a comment here, saying why it is impossible?